File: hCPPLibrary/Network/Buffer/BufferReader.cpp
#include "BufferReader.h"
#include "Log.h"

namespace hlib::net
{
	BufferReader::BufferReader() : buffer_(MAX_BUFFER_SIZE* BUFFER_SPARE)
	{
	}

	bool BufferReader::Write(const void* data, size_t writeSize)
	{
		if (GetRemainSize() < writeSize)
			return false;

		memcpy(buffer_.data() + writeIdx_, data, writeSize);
		writeIdx_ += writeSize;
		return true;
	}

	bool BufferReader::CommitRead(size_t readSize)
	{
		if (GetDataSize() < readSize)
			return false;

		readIdx_ += readSize;
		return true;
	}

	void BufferReader::Cleanup()
	{
		auto dataSize = GetDataSize();
		if (dataSize == 0)
		{
			writeIdx_ = 0;
			readIdx_ = 0;
			return;
		}

		if (GetRemainSize() < MAX_BUFFER_SIZE)
		{
			memmove(buffer_.data(), buffer_.data() + readIdx_, dataSize);
			readIdx_ = 0;
			writeIdx_ = dataSize;
		}
	}

}

File: hCSharpLibrary/Network/Buffer/BufferReader.cs
namespace hCSharpLibrary.Network;

internal class BufferReader
{
	const int BUFFER_SPARE = 3;

	private byte[] _buffer;
	private int _readIdx = 0;
	private int _writeIdx = 0;

	public int DataSize => _writeIdx - _readIdx;
	public int RemainSize => _buffer.Length - _writeIdx;

	public Span<byte> DataSpan => _buffer.AsSpan()[_readIdx.._writeIdx];
	public Memory<byte> DataMemory => _buffer.AsMemory()[_readIdx.._writeIdx];
	public Span<byte> FreeSpan => _buffer.AsSpan()[_writeIdx.._buffer.Length];

	public BufferReader(int bufferSize)
	{
		_buffer = new byte[bufferSize * BUFFER_SPARE];
	}

	public bool Write(Span<byte> buffer)
	{
		if (buffer.Length > RemainSize)
		{
			return false;
		}

		buffer.CopyTo(FreeSpan);
		_writeIdx += buffer.Length;
		return true;
	}

	public bool CommitRead(int size)
	{
		if (size > DataSize)
		{
			return false;
		}

		_readIdx += size;
		return true;
	}

	public void Cleanup()
	{
		if (DataSize == 0)
		{
			_readIdx = 0;
			_writeIdx = 0;
			return;
		}

		DataSpan.CopyTo(_buffer.AsSpan());

		_writeIdx = DataSize;
		_readIdx = 0;
	}
}
